operators/constellation-node-operator/internal/executor/executor.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+// Package executor contains a task executor / scheduler for the constellation node operator.
+// It is used to execute tasks (outside of the k8s specific operator controllers) with regular intervals and
+// based of external triggers.
+package executor
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	defaultPollingFrequency = 15 * time.Minute
+	// rateLimiterItem is the key used to rate limit the reconciliation loop.
+	// since we don't have a reconcile request, we use a constant key.
+	rateLimiterItem = "reconcile"
+)
+
+// Controller is a type with a reconcile method.
+// It is modeled after the controller-runtime reconcile method,
+// but reconciles on external resources instead of k8s resources.
+type Controller interface {
+	Reconcile(ctx context.Context) (Result, error)
+}
+
+// Executor is a task executor / scheduler.
+// It will call the reconcile method of the given controller with a regular interval
+// or when triggered externally.
+type Executor struct {
+	mux     sync.Mutex
+	running bool
+
+	// controller is the controller to be reconciled.
+	controller Controller
+
+	// pollingFrequency is the default frequency with which the controller is reconciled
+	// if no external trigger is received and no requeue is requested by the controller.
+	pollingFrequency time.Duration
+	// rateLimiter is used to rate limit the reconciliation loop.
+	rateLimiter RateLimiter
+	// externalTrigger is used to trigger a reconciliation immediately from the outside.
+	// multiple triggers in a short time will be coalesced into one externalTrigger.
+	externalTrigger chan struct{}
+	// stop is used to stop the reconciliation loop.
+	stop chan struct{}
+}
+
+// New creates a new Executor.
+func New(controller Controller, cfg Config) *Executor {
+	cfg.applyDefaults()
+	return &Executor{
+		controller:       controller,
+		pollingFrequency: cfg.PollingFrequency,
+		rateLimiter:      cfg.RateLimiter,
+	}
+}
+
+// StopWaitFn is a function that can be called to stop the executor and wait for it to stop.
+type StopWaitFn func()
+
+// Start starts the executor in a separate go routine.
+// Call Stop to stop the executor.
+func (e *Executor) Start(ctx context.Context) StopWaitFn {
+	wg := &sync.WaitGroup{}
+	logr := log.FromContext(ctx)
+	stopWait := func() {
+		defer wg.Wait()
+		e.Stop()
+	}
+
+	e.mux.Lock()
+	defer e.mux.Unlock()
+	if e.running {
+		return stopWait
+	}
+	e.running = true
```

Comment:
The `e.mux` looks to me like it just guards the `e.running` boolean. Should we rename `mux` to `runnningMux` or even use an atomic boolean?

Reply:
Atomic boolean is the clearest